apps/src/netsim/NetSimSendWidget.js

```diff
@@ -54,24 +86,123 @@ NetSimSendWidget.createWithin = function (element, connection) {
 NetSimSendWidget.prototype.bindElements_ = function () {
   this.rootDiv_ = $('#netsim_send_widget');
   this.toAddressTextbox_ = this.rootDiv_.find('#to_address');
-  this.payloadTextbox_ = this.rootDiv_.find('#payload');
+  this.toAddressTextbox_.bind('keyPress', function (event) {
+    // Don't block control combinations (copy, paste, etc);
+    if (event.metaKey || event.ctrlKey) {
+      return true;
+    }
+
+    // Prevent non-digit characters
+    var charCode = !event.charCode ? event.which : event.charCode;
+    if (charCode >= 32 && charCode <= 126) {
+      var key = String.fromCharCode(!event.charCode ? event.which : event.charCode);
+
+      if (!/\d/.test(key)) {
+        event.preventDefault();
+        return false;
+      }
+    }
+  });
+  this.toAddressTextbox_.change(this.onToAddressChange_.bind(this));
+
+  this.fromAddressTextbox_ = this.rootDiv_.find('#from_address');
+
+  this.binaryPayloadTextbox_ = this.rootDiv_.find('#binary_payload');
+  this.binaryPayloadTextbox_.bind('keypress', function (event) {
+    // Don't block control combinations (copy, paste, etc);
+    if (event.metaKey || event.ctrlKey) {
+      return true;
+    }
+
+    // Prevent VISIBLE characters that aren't on our whitelist
+    var charCode = !event.charCode ? event.which : event.charCode;
+    if (charCode >= 32 && charCode <= 126) {
+      var key = String.fromCharCode(!event.charCode ? event.which : event.charCode);
+
+      if (!/[01]/.test(key)) {
+        event.preventDefault();
+        return false;
+      }
+    }
+  }.bind(this));
```

Comment:
I don't love the fact that we're calling bind twice here, and it means two different things, but seems to be unavoidable. Was a little surprised that jquery's bind doesn't allow you to give it a context.